File: sdks/nodejs-native/examples/e2e_smoke.mjs
/**
 * EdgeCrab Node.js SDK — E2E Smoke Test
 *
 * Exercises the core Node.js SDK API against local Ollama.
 *
 * Run:
 *   node examples/e2e_smoke.mjs
 */
import { Agent, ModelCatalog, Session } from '../index.js';

const MODEL = 'ollama/gemma4:latest';
const CORE_FEATURES = [
  'ModelCatalog.providerIds',
  'ModelCatalog.modelsForProvider',
  'ModelCatalog.contextWindow',
  'ModelCatalog.pricing',
  'ModelCatalog.flatCatalog',
  'ModelCatalog.defaultModelFor',
  'ModelCatalog.estimateCost',
  'Agent.chat',
  'Agent.run',
  'Agent.runConversation',
  'Agent.stream',
  'Agent.getHistory',
  'Agent.newSession',
  'Agent.chatInCwd',
  'Agent.fork',
  'Agent.batch',
  'Agent.model',
  'Agent.setModel',
  'Agent.sessionSnapshot',
  'Agent.export',
  'Agent.toolNames',
  'Agent.toolsetSummary',
  'Agent.compress',
  'Agent.searchSessions',
  'Agent.listSessions',
  'Session.getMessages',
  'Session.renameSession',
  'Session.pruneSessions',
  'Session.stats',
  'MemoryManager.readWrite',
];
const covered = new Set();

function ok(label) {
  console.log(`  ✓ ${label}`);
}

function section(title) {
  console.log(`\n── ${title} ──`);
}

function mark(feature, label = feature) {
  covered.add(feature);
  ok(label);
}

function expectContains(text, expected) {
  const actual = String(text).trim().toUpperCase();
  if (!actual.includes(expected.toUpperCase())) {
    throw new Error(`Expected ${JSON.stringify(text)} to include ${expected}`);
  }
}

async function main() {
  console.log('EdgeCrab Node.js SDK — E2E Smoke Test');
  console.log(`Model: ${MODEL}`);
  console.log('═══════════════════════════════════════');

  section('1. ModelCatalog');
  const catalog = new ModelCatalog();
  const providers = catalog.providerIds();
  if (!providers.includes('ollama')) {
    throw new Error(`ollama missing from providers: ${providers}`);
  }
  mark('ModelCatalog.providerIds', `${providers.length} providers in catalog`);

  const models = catalog.modelsForProvider('openai');
  if (!Array.isArray(models) || models.length === 0) throw new Error('modelsForProvider empty');
  mark('ModelCatalog.modelsForProvider', `modelsForProvider('openai') → ${models.length} models`);

  const ctxWindow = catalog.contextWindow('openai', 'gpt-4o');
  mark('ModelCatalog.contextWindow', `contextWindow → ${ctxWindow}`);

  const pricing = catalog.pricing('openai', 'gpt-4o');
  mark('ModelCatalog.pricing', `pricing → ${JSON.stringify(pricing)}`);

  const flatCatalog = catalog.flatCatalog();
  if (!Array.isArray(flatCatalog) || flatCatalog.length === 0) throw new Error('flatCatalog empty');
  mark('ModelCatalog.flatCatalog', `flatCatalog → ${flatCatalog.length} total models`);

  const defaultModel = catalog.defaultModelFor('openai');
  if (!defaultModel) throw new Error('defaultModelFor returned empty value');
  mark('ModelCatalog.defaultModelFor', `defaultModelFor('openai') → ${defaultModel}`);

  const estimate = catalog.estimateCost('openai', 'gpt-4o', 1000, 200);
  mark('ModelCatalog.estimateCost', `estimateCost → ${JSON.stringify(estimate)}`);

  section('2. Core agent methods');
  const agent = new Agent({ model: MODEL, maxIterations: 3, quietMode: true,
    skipContextFiles: true, skipMemory: true });

  const reply = await agent.chat('Reply with exactly: PONG');
  expectContains(reply, 'PONG');
  mark('Agent.chat', `chat → ${reply.slice(0, 60).replace(/\n/g, ' ')}`);

  const run = await agent.run('Reply with exactly: OK');
  expectContains(run.response, 'OK');
  mark('Agent.run', `run → ${run.response.slice(0, 40).replace(/\n/g, ' ')}`);

  const conv = await agent.runConversation(
    'Reply with exactly: CONV_OK',
    'You are terse. Obey exactly.',
    ['Previous note: keep replies short.'],
  );
  expectContains(conv.response, 'CONV_OK');
  mark('Agent.runConversation', `runConversation → ${conv.response.slice(0, 40).replace(/\n/g, ' ')}`);

  await agent.setReasoningEffort('low');
  await agent.setStreaming(true);
  const events = await agent.stream('Reply with exactly: STREAM_OK');
  const streamed = events.filter(e => e.eventType === 'token').map(e => e.data).join('');
  expectContains(streamed || JSON.stringify(events), 'STREAM_OK');
  mark('Agent.stream', `stream → ${String(streamed).slice(0, 40).replace(/\n/g, ' ')}`);

  const history = await agent.getHistory();
  if (!history || history.length === 0) throw new Error('history empty after usage');
  mark('Agent.getHistory', `${history.length} messages in history`);

  const cwdReply = await agent.chatInCwd('Reply with exactly: CWD_OK', process.cwd());
  expectContains(cwdReply, 'CWD_OK');
  mark('Agent.chatInCwd', `chatInCwd → ${cwdReply.slice(0, 40).replace(/\n/g, ' ')}`);

  section('3. Batch, fork, and session state');
  const forked = await agent.fork();
  const [forkA, forkB] = await Promise.all([
    forked.chat('Reply with exactly: FORK_A'),
    agent.chat('Reply with exactly: FORK_B'),
  ]);
  expectContains(forkA, 'FORK_A');
  expectContains(forkB, 'FORK_B');
  mark('Agent.fork', `fork replies → ${forkA.slice(0, 10)} / ${forkB.slice(0, 10)}`);

  const batch = await agent.batch(['Reply with exactly: YES', 'Reply with exactly: NO']);
  if (batch.length !== 2) throw new Error(`batch returned ${batch.length} items`);
  expectContains(batch[0], 'YES');
  expectContains(batch[1], 'NO');
  mark('Agent.batch', `batch → ${batch.map(v => String(v).trim()).join(', ')}`);

  const snapshot = await agent.sessionSnapshot();
  if (!snapshot || typeof snapshot !== 'object') throw new Error('sessionSnapshot invalid');
  mark('Agent.sessionSnapshot', `sessionSnapshot → ${snapshot.messageCount ?? snapshot.message_count ?? 'ok'} messages`);

  const exported = await agent.export();
  if (!exported || !Array.isArray(exported.messages)) throw new Error('export invalid');
  mark('Agent.export', `export → ${exported.messages.length} messages`);

  section('4. Hot-swap, reset, and metadata');
  const modelBefore = await agent.model;
  if (!modelBefore) throw new Error('model getter empty');
  mark('Agent.model', `model getter → ${modelBefore}`);

  await agent.setModel(MODEL);
  mark('Agent.setModel', `setModel → ${MODEL}`);

  const toolNames = await agent.toolNames();
  if (!Array.isArray(toolNames)) throw new Error('toolNames returned a non-array value');
  mark('Agent.toolNames', `toolNames → ${toolNames.length} tools`);

  const toolsetSummary = await agent.toolsetSummary();
  if (!Array.isArray(toolsetSummary)) throw new Error('toolsetSummary returned a non-array value');
  mark('Agent.toolsetSummary', `toolsetSummary → ${toolsetSummary.length} groups`);

  const sid = await agent.sessionId;
  if (!sid) throw new Error('sessionId empty');
  ok(`sessionId → ${sid.slice(0, 12)}…`);

  await agent.newSession();
  const resetHistory = await agent.getHistory();
  if (resetHistory.length !== 0) throw new Error('newSession did not clear history');
  mark('Agent.newSession', 'newSession cleared history');

  section('5. Session store and memory');
  const hits = agent.searchSessions('PONG', 5);
  mark('Agent.searchSessions', `searchSessions → ${hits.length} hits`);

  const sessions = agent.listSessions(5);
  mark('Agent.listSessions', `listSessions → ${sessions.length} sessions`);

  const sessionStore = new Session();
  const standaloneMessages = sessionStore.getMessages(sid);
  if (!Array.isArray(standaloneMessages)) throw new Error('getMessages failed');
  mark('Session.getMessages', `getMessages → ${standaloneMessages.length} messages`);

  sessionStore.renameSession(sid, 'Node SDK E2E');
  mark('Session.renameSession', 'renameSession applied');

  const pruned = sessionStore.pruneSessions(36500);
  mark('Session.pruneSessions', `pruneSessions → ${pruned} deleted`);

  const stats = sessionStore.stats();
  mark('Session.stats', `session stats → ${stats.totalSessions} sessions / ${stats.totalMessages} messages`);

  const mem = agent.memory;
  const token = `node-sdk-${Date.now()}`;
  await mem.write('memory', token);
  const memContent = await mem.read('memory');
  if (!memContent.includes(token)) throw new Error('memory write/read failed');
  const entries = await mem.entries('memory');
  if (!Array.isArray(entries)) throw new Error('memory entries failed');
  const removed = await mem.remove('memory', token);
  if (!removed) throw new Error('memory remove failed');
  mark('MemoryManager.readWrite', `memory round-trip → ${entries.length} entries`);

  await agent.compress();
  mark('Agent.compress', 'compress completed');

  const pct = (covered.size / CORE_FEATURES.length) * 100;
  console.log(`\nCore API coverage: ${covered.size}/${CORE_FEATURES.length} (${pct.toFixed(1)}%)`);
  if (pct < 80) {
    throw new Error(`Coverage below target: ${pct.toFixed(1)}%`);
  }

  console.log('\n═══════════════════════════════════════');
  console.log('E2E smoke test PASSED ✓');
  console.log('Node.js SDK coverage target PASSED ✓');
}

main().catch(err => {
  console.error('E2E FAILED:', err);
  process.exit(1);
});

File: sdks/python/examples/e2e_smoke.py
"""
EdgeCrab Python SDK — E2E Smoke Test

Exercises the core Python SDK API against local Ollama.
Run with:
    python e2e_smoke.py
"""
import asyncio
import sys
from pathlib import Path

MODEL = "ollama/gemma4:latest"
CORE_FEATURES = [
    "ModelCatalog.provider_ids",
    "ModelCatalog.pricing",
    "ModelCatalog.flat_catalog",
    "ModelCatalog.default_model_for",
    "ModelCatalog.estimate_cost",
    "AsyncAgent.chat",
    "AsyncAgent.run",
    "AsyncAgent.run_conversation",
    "AsyncAgent.stream",
    "AsyncAgent.history",
    "AsyncAgent.session_id",
    "AsyncAgent.chat_in_cwd",
    "AsyncAgent.fork",
    "AsyncAgent.batch",
    "AsyncAgent.set_model",
    "AsyncAgent.session_snapshot",
    "AsyncAgent.export",
    "AsyncAgent.tool_names",
    "AsyncAgent.toolset_summary",
    "AsyncAgent.compress",
    "AsyncAgent.new_session",
    "AsyncAgent.search_sessions",
    "AsyncAgent.list_sessions",
    "Session.get_messages",
    "Session.rename_session",
    "Session.prune_sessions",
    "Session.stats",
    "AsyncMemoryManager.read_write",
]
covered: set[str] = set()


def ok(label: str) -> None:
    print(f"  ✓ {label}")


def section(title: str) -> None:
    print(f"\n── {title} ──")


def mark(feature: str, label: str) -> None:
    covered.add(feature)
    ok(label)


def expect_contains(text: str | None, expected: str) -> None:
    actual = str(text or "").strip().upper()
    if expected.upper() not in actual:
        raise AssertionError(f"Expected {text!r} to include {expected!r}")


async def main() -> None:
    print("EdgeCrab Python SDK — E2E Smoke Test")
    print(f"Model: {MODEL}")
    print("═══════════════════════════════════════")

    try:
        from edgecrab import AsyncAgent, Session
        from edgecrab._native import ModelCatalog
    except ImportError as e:
        print(f"FATAL: cannot import edgecrab: {e}")
        sys.exit(1)

    section("1. ModelCatalog")
    catalog = ModelCatalog()
    providers = catalog.provider_ids()
    assert providers, "No providers in catalog"
    assert "ollama" in providers, f"ollama missing from {providers}"
    mark("ModelCatalog.provider_ids", f"{len(providers)} providers in catalog")

    pricing = catalog.pricing("openai", "gpt-4o")
    mark("ModelCatalog.pricing", f"pricing lookup → {pricing}")

    flat = catalog.flat_catalog()
    assert flat, "flat_catalog empty"
    mark("ModelCatalog.flat_catalog", f"flat_catalog → {len(flat)} models")

    default_model = catalog.default_model_for("openai")
    assert default_model, "default_model_for returned None"
    mark("ModelCatalog.default_model_for", f"default_model_for('openai') → {default_model}")

    est = catalog.estimate_cost("openai", "gpt-4o", 1000, 200)
    mark("ModelCatalog.estimate_cost", f"cost estimate → {est}")

    section("2. Core agent methods")
    agent = AsyncAgent(MODEL, max_iterations=3, quiet_mode=True,
                       skip_context_files=True, skip_memory=True)
    reply = await agent.chat("Reply with exactly: PONG")
    expect_contains(reply, "PONG")
    mark("AsyncAgent.chat", f"chat → {reply[:60].replace(chr(10), ' ')}")

    result = await agent.run("Reply with exactly: OK")
    assert result.response, "response empty"
    expect_contains(result.response, "OK")
    mark("AsyncAgent.run", f"run → {result.response[:40].replace(chr(10), ' ')}")

    conv = await agent.run_conversation(
        "Reply with exactly: CONV_OK",
        system="You are terse. Obey exactly.",
        history=["Earlier note: keep replies short."],
    )
    expect_contains(conv.response, "CONV_OK")
    mark("AsyncAgent.run_conversation", f"run_conversation → {conv.response[:40].replace(chr(10), ' ')}")

    await agent.set_reasoning_effort("low")
    await agent.set_streaming(True)
    stream_parts: list[str] = []
    async for event in agent.stream("Reply with exactly: STREAM_OK"):
        if event.event_type == "token":
            stream_parts.append(event.data)
    streamed = "".join(stream_parts)
    expect_contains(streamed or str(stream_parts), "STREAM_OK")
    mark("AsyncAgent.stream", f"stream → {streamed[:40].replace(chr(10), ' ')}")

    msgs = agent.history
    assert msgs, "history empty after run()"
    mark("AsyncAgent.history", f"{len(msgs)} messages in history")

    sid = agent.session_id
    assert sid, "session_id is None"
    mark("AsyncAgent.session_id", f"session_id = {sid[:12]}…")

    cwd_reply = await agent.chat_in_cwd("Reply with exactly: CWD_OK", str(Path.cwd()))
    expect_contains(cwd_reply, "CWD_OK")
    mark("AsyncAgent.chat_in_cwd", f"chat_in_cwd → {cwd_reply[:40].replace(chr(10), ' ')}")

    section("3. Batch, fork, and state")
    fork_a = await agent.fork()
    fork_b = await agent.fork()
    ra, rb = await asyncio.gather(
        fork_a.chat("Reply with exactly: FORK_A"),
        fork_b.chat("Reply with exactly: FORK_B"),
    )
    expect_contains(ra, "FORK_A")
    expect_contains(rb, "FORK_B")
    mark("AsyncAgent.fork", f"fork replies → {ra[:10]} / {rb[:10]}")

    results = await agent.batch(["Reply with exactly: YES", "Reply with exactly: NO"])
    assert len(results) == 2, f"batch returned {len(results)} items"
    expect_contains(results[0], "YES")
    expect_contains(results[1], "NO")
    mark("AsyncAgent.batch", f"batch → {[str(r).strip() for r in results]}")

    snapshot = await agent.session_snapshot()
    assert isinstance(snapshot, dict), "session_snapshot invalid"
    mark("AsyncAgent.session_snapshot", f"session_snapshot → {snapshot.get('message_count', 'ok')} messages")

    exported = await agent.export()
    assert isinstance(exported, dict) and isinstance(exported.get("messages"), list), "export invalid"
    mark("AsyncAgent.export", f"export → {len(exported['messages'])} messages")

    section("4. Metadata and persistence")
    await agent.set_model(MODEL)
    mark("AsyncAgent.set_model", f"set_model → {agent.model}")

    tool_names = await agent.tool_names()
    assert isinstance(tool_names, list), "tool_names returned a non-list value"
    mark("AsyncAgent.tool_names", f"tool_names → {len(tool_names)} tools")

    toolset_summary = await agent.toolset_summary()
    assert isinstance(toolset_summary, list), "toolset_summary returned a non-list value"
    mark("AsyncAgent.toolset_summary", f"toolset_summary → {len(toolset_summary)} groups")

    hits = agent.search_sessions("PONG", 5)
    mark("AsyncAgent.search_sessions", f"search_sessions → {len(hits)} hits")

    sessions = agent.list_sessions(5)
    mark("AsyncAgent.list_sessions", f"list_sessions → {len(sessions)} sessions")

    db = Session()
    messages = db.get_messages(sid)
    assert isinstance(messages, list), "Session.get_messages failed"
    mark("Session.get_messages", f"Session.get_messages → {len(messages)} messages")

    db.rename_session(sid, "Python SDK E2E")
    mark("Session.rename_session", "rename_session applied")

    pruned = db.prune_sessions(36500)
    mark("Session.prune_sessions", f"prune_sessions → {pruned} deleted")

    stats = db.stats()
    mark("Session.stats", f"session stats → {stats.total_sessions} sessions / {stats.total_messages} messages")

    mem = agent.memory
    token = f"python-sdk-{int(asyncio.get_running_loop().time() * 1000)}"
    await mem.write("memory", token)
    content = await mem.read("memory")
    assert token in content, "memory write/read failed"
    entries = await mem.entries("memory")
    removed = await mem.remove("memory", token)
    assert removed, "memory remove failed"
    mark("AsyncMemoryManager.read_write", f"memory round-trip → {len(entries)} entries")

    await agent.compress()
    mark("AsyncAgent.compress", "compress completed")

    await agent.new_session()
    assert agent.history == [], "new_session did not clear history"
    mark("AsyncAgent.new_session", "new_session cleared history")

    pct = (len(covered) / len(CORE_FEATURES)) * 100
    print(f"\nCore API coverage: {len(covered)}/{len(CORE_FEATURES)} ({pct:.1f}%)")
    if pct < 80:
        raise AssertionError(f"Coverage below target: {pct:.1f}%")

    print("\n═══════════════════════════════════════")
    print("E2E smoke test PASSED ✓")
    print("Python SDK coverage target PASSED ✓")


if __name__ == "__main__":
    asyncio.run(main())
